File: EmojibleIOS/EmojibleIOS/Interpreter/AST.swift
//
//  AST.swift
//  EmojibleIOS
//
//  Created by Haldun Balim on 3.12.2020.
//

import Foundation

fileprivate let runScreen = RunCodeCoordinator.getInstance().runScreen!
fileprivate var isCancelled: Bool { Interpreter.getInstance().isCancelled }

enum InterpreterErrors: Error {
    case GenericError(str:String)
}

enum EmojibleErrors: Error {
    case C1
}

class AST{
    func visit() throws -> Any?{
        return nil
    }
    
    func safeVisit() throws -> Any?{
        if isCancelled{
            return nil
        }
        do{
            return try self.visit()
        }catch{
            throw error
        }
    }
}
    
class BinOpNode:AST{
    var left:AST
    var op:Token
    var token:Token
    var right:AST
    
    init(left:AST, op:Token, right:AST){
        self.left = left
        self.token = op
        self.op = op
        self.right = right
        
    }
    
    override func visit() throws -> Any?{
        do{
            let left = try self.left.safeVisit()
            let right = try self.right.safeVisit()
            if let left_val = left  as? Int, let right_val = right as? Int{
                return visitInt(left:left_val , right:right_val)
            }else if let left_val = left as? Bool, let right_val = right as? Bool{
                return visitBool(left:Bool(left_val), right:Bool(right_val))
            }else if let left_val = left as? NSNumber, let right_val = right as? NSNumber{
                return visitFloat(left:Float(truncating: left_val), right:Float(truncating: right_val))
            }
            else{
                throw InterpreterErrors.GenericError(str: "Expected Number or Boolean for binary operator at line \(String(describing:self.token.lineNumber))")
            }
        }
        catch{
            throw error
        }
    }
    
    func visitBool(left:Bool,right:Bool) -> Any?{
        if op.type == TokenType.AND{
            return left && right
        }else if op.type == TokenType.OR{
            return left || right
        }
        return nil
    }
    
    func visitFloat(left:Float,right:Float) -> Any?{
        if op.type == TokenType.PLUS{
            return left + right
        }else if op.type == TokenType.MINUS{
            return left - right
        }else if op.type == TokenType.MUL{
            return left * right
        }else if op.type == TokenType.FLOAT_DIV{
            return left/right
        }else if op.type == TokenType.EQUALS{
            return left == right
        }else if op.type == TokenType.LESSER_EQ{
            return left <= right
        }else if op.type == TokenType.GREATER_EQ{
            return left >= right
        }else if op.type == TokenType.LESSER{
            return left < right
        }else if op.type == TokenType.GREATER{
            return left > right
        }
        return nil
    }
    
    func visitInt(left:Int,right:Int) -> Any?{
        let result = visitFloat(left: Float(left), right: Float(right))
        if result is Bool{
            return result
        }
        
        let resultInt:Int = Int((result as! Float).rounded(.down))
        
        
        
        return resultInt
    }
    
}

class TouchedNode:AST{
    var token: Token
    
    init(token:Token){
        self.token = token
        
    }
    
    override func visit() -> Any? {
        return TouchSensor.getInstance().isTouched()
    }
}

class FlippedNode:AST{
    var token: Token
    
    init(token:Token){
        self.token = token
        
    }
    
    override func visit() -> Any? {
        return FlipSensor.getInstance().isFlipped()
    }
}

class ValueNode:AST{
    var token: Token
    var value: Any
    
    init(token:Token){
        self.token = token
        self.value = token.value
        
    }
    
    override func visit() -> Any? {
        return value
    }
}

class NumNode:ValueNode{
    
}

class BoolNode:ValueNode{
    
}


class ColorNode:ValueNode{
    
}

class GetNumericUserInputNode:AST{
    var op:TokenType
    
    init(op: TokenType){
        self.op = op
        
    }
    
    override func visit() throws ->Any?{
        runScreen.performSelector(onMainThread: #selector(runScreen.textInputRequested), with: nil, waitUntilDone: false)
        Interpreter.getInstance().inputSemaphore.wait()
        
        if isCancelled{
            return nil
        }
        
        let input = runScreen.input!
        if let inputInt = Int(input){
            return inputInt
        }else if let inputFloat = Float(input){
            return inputFloat
        }else{
            throw InterpreterErrors.GenericError(str: "\(input) be converted to number")
        }
    }
    
}

class SetScreenColorNode:AST{
    var color: ColorNode
    var op: TokenType
    
    init(op:TokenType, color:ColorNode){
        self.op = op
        self.color = color
        
    }
    
    override func visit()->Any?{
        let backgroundColor = color.token.value as? String
        runScreen.backgroundColor = backgroundColor
        runScreen.performSelector(onMainThread: #selector(runScreen.changeBackgroundColor), with: nil, waitUntilDone: false)
        return nil
    }
}

class DisplayNode:AST{
    var token:Token
    var op:Token
    var expr:AST
    init(op:Token, expr:AST){
        self.token = op
        self.op = op
        self.expr = expr
        
    }
    
    override func visit() throws -> Any? {
        do{
            let expr_res = try expr.safeVisit()
            if let str = expr_res as? String {
                if str.contains(".m4a") && FileSystemManager.getInstance().fileExists(filename: str){
                    AudioPlayer.getInstance().playAudio(filename: str)
                }else{
                    runScreen.outText = str
                    runScreen.performSelector(onMainThread: #selector(runScreen.changeLabelText), with: nil, waitUntilDone: false)
                }
            }else if let str = expr_res as? Int {
                runScreen.outText = String(describing: str)
                runScreen.performSelector(onMainThread: #selector(runScreen.changeLabelText), with: nil, waitUntilDone: false)
            }else if let str = expr_res as? Float {
                runScreen.outText = String(describing: str)
                runScreen.performSelector(onMainThread: #selector(runScreen.changeLabelText), with: nil, waitUntilDone: false)
            }
            else if let str = expr_res as? Bool {
                runScreen.outText = String(describing: str)
                runScreen.performSelector(onMainThread: #selector(runScreen.changeLabelText), with: nil, waitUntilDone: false)
            }
            return nil
        }catch{
            throw error
        }
    }
        
}

class TTSNode:AST{
    var token:Token
    var op:Token
    var expr:AST
    init(op:Token, expr:AST){
        self.token = op
        self.op = op
        self.expr = expr
        
    }
    
    override func visit() throws -> Any? {
        do{
            let expr_res = try expr.safeVisit()
            if let str = expr_res  as? String {
                TextToSpeech.getInstance().convertTextToSpeech(text: str, language: "English")
            }
            return nil
        }catch{
            throw error
        }
    }
        
}



class UnaryOpNode:AST{
    var token:Token
    var op:Token
    var expr:AST
    init(op:Token, expr:AST){
        self.token = op
        self.op = op
        self.expr = expr
        
    }
    
    override func visit() throws -> Any? {
        do{
            let expr_res = try expr.safeVisit()
            if let num = expr_res  as? Int {
                if op.type == TokenType.PLUS{
                    return +num
                }else if op.type == TokenType.MINUS{
                    return -num
                }
            }
            if let num = expr_res as? Float {
                if op.type == TokenType.PLUS{
                    return +num
                }else if op.type == TokenType.MINUS{
                    return -num
                }
            }
            return nil
        }
        catch{
            throw error
        }
    }
}

class CompoundNode:AST{
    var children: [AST]
    init(children:[AST]?=nil){
        self.children = children == nil ? []:children!
    }
    
    override func visit() throws-> Any? {
        do{
            for child in children{
                _ = try child.safeVisit()
                if isCancelled{
                    return nil
                }
            }
            return nil
        }
        catch{
            throw error
        }
    }
}

class AssignNode:AST{
    var left: VarNode
    var op:Token
    var token:Token
    var right:AST
    var memory:Memory
    
    init(left:VarNode, op:Token, right:AST, memory:Memory){
        self.left = left
        self.token = op
        self.op = op
        self.right = right
        self.memory = memory
        
    }
    
    override func visit() throws -> Any? {
        do{
            let var_name = left.value
            let var_value = try right.safeVisit()
            if isCancelled{
                return nil
            }
            memory.addAssignment(assignment: AssignmentModel(identifier: var_name as! String, value: var_value!))
            return nil
        }catch{
            throw error
        }
    }
}

class GetRandomNumberNode:AST{
    var op:Token
    var token:Token
    var lower_bound:AST
    var upper_bound:AST
    init(op:Token, lower_bound:AST, upper_bound:AST){
        self.token = op
        self.op = op
        self.lower_bound = lower_bound
        self.upper_bound = upper_bound
        
    }
    
    override func visit() throws -> Any? {
        
        do{
            guard let lower_bound = try self.lower_bound.safeVisit() as? Int else{
                throw InterpreterErrors.GenericError(str:"Random number cannot be called with non integer lower bound")
            }
            guard let upper_bound = try self.upper_bound.safeVisit() as? Int else{
                throw InterpreterErrors.GenericError(str:"Random number cannot be called with non integer upper bound")
            }
            if lower_bound >= upper_bound{
                throw InterpreterErrors.GenericError(str:"lower bound cannot be greater than upper bound")
            }
            return Int.random(in: lower_bound ..< upper_bound)
        }catch{
            throw error
        }
    }
}

class IfNode:AST{
    var op:TokenType
    var bool_statement:AST
    var true_statement:BlockNode
    var false_statement:BlockNode
    init(op:TokenType, bool_statement:AST, true_statement:BlockNode, false_statement:BlockNode?=nil){
        self.op = op
        self.bool_statement = bool_statement
        self.true_statement = true_statement
        if let false_st = false_statement{
            self.false_statement = false_st
        }else{
            let children: [AST] = [NoOpNode()]
            self.false_statement = BlockNode(compound_statement: CompoundNode(children:children))
        }
    }
    
    override func visit() throws -> Any? {
        do{
            guard let bool = try bool_statement.safeVisit() as? Bool else {
                throw InterpreterErrors.GenericError(str:"If statement must have boolean first argument")
                return nil
            }
            if bool{
                return try true_statement.safeVisit()
            }else{
                return try false_statement.safeVisit()
            }
        }catch{
            throw error
        }
    }
}

class ForNode:AST{
    var op:TokenType
    var times:AST
    var body:BlockNode
    var hasPlayerInBody: Bool
    init(op:TokenType, times:AST, body:BlockNode, hasPlayerInBody:Bool){
        self.op = op
        self.times = times
        self.body = body
        self.hasPlayerInBody = hasPlayerInBody
        
    }
    
    override func visit() throws -> Any? {
        do{
            guard let times = try self.times.safeVisit() as? Int else{
                throw InterpreterErrors.GenericError(str:"For statement must have integer times argument")
            }
            for _ in 0..<times{
                _ = try body.safeVisit()
                if isCancelled{
                    return nil
                }
                if hasPlayerInBody{
                    usleep(useconds_t(Constants.SLEEP_DURATION_IN_FOR))
                }
            }
            return nil
        }
        catch{
            throw error
        }
    }
}

class WhileNode:AST{
    var op:TokenType
    var bool_statement:AST
    var body:BlockNode
    var hasSensorInBody: Bool
    init(op:TokenType, bool_statement:AST, body:BlockNode, hasSensorInBody:Bool){
        self.op = op
        self.bool_statement = bool_statement
        self.body = body
        self.hasSensorInBody = hasSensorInBody
        
    }
    
    override func visit() throws -> Any? {
        do{
            while true{
                guard let bool_cond = try bool_statement.safeVisit() as? Bool else{
                    return nil

                }
                if bool_cond{
                    break
                }
                if isCancelled{
                    return nil
                }
                _ = try body.safeVisit()
                if self.hasSensorInBody{
                    usleep(useconds_t(Constants.SLEEP_DURATION_IN_WHILE))
                }
            }
            return nil
        }catch{
            throw error
        }
    }
}

class VarNode:AST{
    var token:Token
    var value:Any
    var memory:Memory
    init(token:Token, memory:Memory){
        self.token = token
        self.value = token.value
        self.memory = memory
    }
    
    override func visit() throws -> Any? {
        
        let var_name = value as! String
        
        for assignment in memory.getAssignments(){
            if assignment.identifier == var_name{
                return assignment.getValue()
            }
        }
        throw InterpreterErrors.GenericError(str:"\(var_name) is an undefined variable")
        
    }
}

class NoOpNode:AST{
}

class ProgramNode:AST{
    var block:BlockNode
    init(block:BlockNode){
        self.block = block
    }
    
    override func visit() throws -> Any? {
        do{
            return try block.safeVisit()
        }catch{
            throw error
        }
    }
}

class BlockNode:AST{
    var compound_statement:CompoundNode
    init(compound_statement:CompoundNode){
        self.compound_statement = compound_statement
    }
    override func visit() throws -> Any? {
        do{
            return try compound_statement.safeVisit()
        }catch{
            throw error
        }
    }
}

File: EmojibleAndroid/app/src/main/java/com/dji/emojibleandroid/interpreter/AST.kt
package com.dji.emojibleandroid.interpreter

import com.dji.emojibleandroid.models.AssignmentModel
import com.dji.emojibleandroid.models.Value
import com.dji.emojibleandroid.models.ValueType
import com.dji.emojibleandroid.utils.CodeScreenUtils

val runScreen = CodeScreenUtils.runScreen
val cancelled = Interpreter.instance.job!!.isCancelled

abstract class AST {
    open suspend fun visit(): Any? {
        return null
    }

    suspend fun safeVisit(): Any? {
        if (cancelled) {
            return null
        }
        return visit()
    }
}

class BinOpNode(private val left: AST, private val op: Token, private val right: AST) : AST() {
    val token: Token = op
    override suspend fun visit(): Any? {
        val leftVal = left.safeVisit() as Value?
        val rightVal = right.safeVisit() as Value?
        println(leftVal?.javaClass?.name)
        if (leftVal == null) {
            throw Exception("Left element of the binary operation is null")
        }
        if (rightVal == null) {
            throw Exception("Right element of the binary operation is null")
        }
        if ((leftVal.type == ValueType.Float) || rightVal.type == ValueType.Float) {
            return visitFloat(leftVal.value as Float, rightVal.value as Float)
        } else if (leftVal.type == ValueType.Integer && rightVal.type == ValueType.Integer) {
            return visitInt(leftVal.value as Int, rightVal.value as Int)
        }
        throw Exception("Unknown type for binary operation")
    }

    private fun visitFloat(leftVal: Float, rightVal: Float): Any? {
        return when (op.type) {
            TokenType.PLUS -> leftVal + rightVal
            TokenType.MINUS -> leftVal - rightVal
            TokenType.MUL -> leftVal * rightVal
            TokenType.FLOAT_DIV -> leftVal / rightVal
            TokenType.EQUALS -> leftVal == rightVal
            TokenType.LESSER_EQ -> leftVal <= rightVal
            TokenType.GREATER_EQ -> leftVal >= rightVal
            TokenType.LESSER -> leftVal < rightVal
            TokenType.GREATER -> leftVal > rightVal
            else -> null
        }
    }

    private fun visitInt(leftVal: Int, rightVal: Int): Any? {
        val result = visitFloat(leftVal.toFloat(), rightVal.toFloat())
        if (result is Boolean)
            return result


        if (result is Float && (result - result.toInt()) == 0.toFloat())
            return result.toInt()

        return result
    }
}

open class ValueNode(private val token: Token) : AST() {
    val value: Any? = token.value
    override suspend fun visit(): Any? {
        return token.value
    }

}

class NumNode(token: Token) : ValueNode(token)
class BoolNode(token: Token) : ValueNode(token)
class ColorNode(token: Token) : ValueNode(token)
class GetNumericUserInputNode(val tokenType: TokenType) : AST() {
    override suspend fun visit(): Any? {
        CodeScreenUtils.runScreen?.userInputRequested()
        Interpreter.instance.inputSemaphore?.acquire()
        if (cancelled)
            return null
        var userInput = CodeScreenUtils.runScreen?.input
        while (userInput == "") {
            Interpreter.instance.inputSemaphore?.release()
            CodeScreenUtils.runScreen?.userInputRequested()
            Interpreter.instance.inputSemaphore?.acquire()
            userInput = CodeScreenUtils.runScreen?.input
        }
        Interpreter.instance.inputSemaphore?.release()
        return when {
            userInput?.toIntOrNull() != null -> {
                userInput.toInt()
            }
            userInput?.toFloatOrNull() != null -> {
                userInput.toFloat()
            }
            else -> {
                null
            }
        }
    }
}

class SetScreenColorNode(val tokenType: TokenType, color: ColorNode) : AST() {
    val color: String = color.value as String
    override suspend fun visit(): Any? {
        CodeScreenUtils.runScreen?.changeBackgroundColor(color)
        return null
    }
}

class UnaryOpNode(private val op: Token, private val expr: AST) : AST() {
    override suspend fun visit(): Any? {
        val num = expr.safeVisit()
        if (num is Int) {
            if (op.type == TokenType.PLUS)
                return +num
            else if (op.type == TokenType.MINUS)
                return -num
        }
        if (num is Float) {
            if (op.type == TokenType.PLUS)
                return +num
            else if (op.type == TokenType.MINUS)
                return -num
        }
        return null
    }
}

class CompoundNode() : AST() {
    val children: MutableList<AST> = mutableListOf()
    override suspend fun visit(): Any? {
        for (child in children) {
            child.safeVisit()
            if (cancelled)
                return null
        }
        return null
    }
}

class AssignNode(
    private val left: VarNode,
    op: Token,
    private val right: AST,
    private val memory: Memory
) : AST() {
    val token: Token = op
    override suspend fun visit(): Any? {
        val varName = left.value
        val varValue = right.visit()
        if (cancelled)
            return null
        varValue?.let { AssignmentModel(varName as CharSequence, it) }?.let {
            memory.addAssignment(
                it
            )
        }
        return null
    }
}

class GetRandomNumberNode(
    token: TokenType,
    private val lowerBound: AST,
    private val upperBound: AST
) : AST() {
    val op: TokenType = token
    override suspend fun visit(): Any {
        val lowerBound = lowerBound.safeVisit()
        val upperBound = upperBound.safeVisit()
        if (lowerBound !is Int && !(lowerBound is Float && lowerBound as Int == lowerBound)) {
            throw Exception("Random number cannot be called with non integer lower bound")
        } else if (upperBound !is Int && !(upperBound is Float && upperBound as Int == upperBound)) {
            throw Exception("Random number cannot be called with non integer upper bound")
        }
        if (lowerBound >= upperBound)
            throw Exception("lowerBound $lowerBound must be lower than upperBound $upperBound")
        return (lowerBound..upperBound).random()
    }
}

class IfNode(
    val op: TokenType,
    private val boolStatement: AST,
    private val trueStatement: BlockNode,
    var falseStatement: AST?
) : AST() {
    init {
        falseStatement = if (falseStatement != null) falseStatement as BlockNode else NoOpNode()
    }

    override suspend fun visit() {
        val boolSt = boolStatement.safeVisit()
        if (boolSt !is Boolean) {
            throw Exception("The statement after for keyword must be a boolean")
        }
        if (boolSt) {
            trueStatement.safeVisit()
        } else {
            falseStatement?.safeVisit()
        }
    }
}

class ForNode(op: TokenType, val times: AST, private val body: BlockNode) : AST() {
    val token: TokenType = op
    override suspend fun visit(): Any? {
        val times = times.safeVisit()
        if (times !is Int || (times is Float && times == times as Int)) {
            throw Exception("$times is not an integer")
        }
        for (i in 0..times) {
            body.safeVisit()
            if (cancelled)
                return null
        }
        return null
    }
}

class WhileNode(op: TokenType, private val boolStatement: AST, private val body: AST) : AST() {
    val token: TokenType = op
    override suspend fun visit() {
        while (true) {
            val boolCond = boolStatement.safeVisit()
            if (boolCond !is Boolean) {
                throw Exception("$boolCond is not a boolean")
            }
            if (boolCond) {
                break
            }
            body.safeVisit()
        }
    }
}

class VarNode(token: Token, private val memory: Memory) : AST() {
    val value: Any? = token.value
    override suspend fun visit(): Any {
        val varName = value as CharSequence
        for (assignment in memory.assignments) {
            if (assignment.identifier == varName) {
                return assignment.value
            }
        }
        throw Exception("$varName is an undefined variable")
    }
}

class NoOpNode() : AST() {
    override suspend fun visit() {}
}

class ProgramNode(private val blockNode: BlockNode) : AST() {
    override suspend fun visit() {
        blockNode.safeVisit()
    }
}

class BlockNode(private val compoundNodeStatement: CompoundNode) : AST() {
    //    var declarations: MutableList<AST> = mutableListOf()
    override suspend fun visit() {
        compoundNodeStatement.safeVisit()
    }
}

